Implementation of an alternate algorithm for reciprocal values

stesser opened the pull request:
I'm still not sure about correct scale values and situations in which bc_num_extend must be called to prepare a BcNum that is to receive the result of an arithmetic operation. This leads to overflow for large values (the algorithm is essentially correct, but variables need to be extended in some cases).

stesser committed:
Updated implementation of Newton-Raphson division algorithm to match upstream changes

stesser committed:
Small optimization of the division function

stesser committed:
Small optimization of the division function

stesser committed:
Merge branch 'base9' of https://github.com/gavinhoward/bc into d9

stesser committed:
Undo change to scale calculation in bc_num_m

I had missed to set scsale values in the parameters passed in by the divide
function and thus got results that had 0 fractional digits.

stesser committed:
Add function to round number to a given number of decimal places

This function will be used in the division algorithm.

stesser committed:
Fix division: Extend variable to required number of decimal places

I had a local change that made bc_num_invert implicily extend this variable,
but noticed with that change reverted that the precision of the calculation
was reduced to BC_BASE_POWER decimals, leading to imprecise division results.

stesser committed:
Replace Newton-Raphson algorithm by Goldschmidt algorithm.

For now, keep the Newton-Raphson algorithm in an #if 0 block for reference and to
allow to compare them in further tests.

I have noticed segmentation faults when dividing very large numbers, which probably
are due to missing calls of bc_num_extend() for destination variables of arithmetic
operations. I'll try to fix the length calculations, but I do not know the exact
semantics of a number of low level calls.

gavinhoward commented:
How does this new algorithm compare to Newton-Raphson performance-wise?

stesser committed:
Merge branch 'base9' of https://github.com/gavinhoward/bc into d9

gavinhoward commented:
Also, would it be possible for you to fix the crashes on both algorithms? The test suite crashes on both. It would also be nice if you could put them in a #if 0 #else #endif set so I can test the changes and see which one is faster. Thank you.

stesser committed:
Optimize bc_num_pow10

stesser committed:
Fix scale parameters and remove bogus bc_num_extend calls

This makes bc complete the tests without crash, but with a result that
has too few digits printed. This will need to be debugged next ...

stesser committed:
Fix position of decimal point when dividing by a very small number.

Now all division and multiplication tests pass.

stesser committed:
Improve format of dumped BcNum values in debug traces

stesser committed:
Fix "power" operator (bc_num_p).

There were 4 places where "rdx" had to be replaced by "scale".
This version passes all tests.

stesser committed:
Mostly fix the sqrt operation.

Results seem to be correct except for the number of decimals calculated and returned.
The tests still fail for that reason, but show results that are near to what is expected.

stesser committed:
Fix and improve the calculation of the initial value for the sqrt algorithm

Introduce and use a function bc_num_int_digits() that counts the number of digits
to the left of the decimalpoint.

stesser committed:
Make more Sqrt tests work

stesser committed:
Make all Sqrt tests work

stesser committed:
Apply existing style to my changes - no functional change

stesser committed:
Fix debug printing of BcNum variables

stesser committed:
Make debug printing of BcNum variables use signed values to make overuns easier to decode

stesser committed:
Fix printing of numbers with obase != 10 (length of fractional part)

stesser committed:
Merge branch 'base9' of https://github.com/gavinhoward/bc into d9

stesser committed:
Revert bc_num_pow10 to a function instead of an array that holds the results.

The implementation can be changed back to a macro that accesses the array
without any further computation, if the function should take non-negligible
time. But the function will be inlined or optimized away in most cases and
thus should not have a significant impact on the run-time ...

stesser committed:
Add back the forward declaration of bc_num_pow10.

stesser committed:
Allow to pass BC_BASE_POWER on the compiler command line

stesser committed:
Fix decoding of BcNum in debug print function

stesser committed:
Fix merge errors in idivision based on Goldschmidt algorithm

stesser committed:
Change order of tests to first test features that later tests depend on

"shift" operations are used within all arithmetic operations
"print" is a pre-requisite of generating the results file for the "parse" test

stesser committed:
Change type of variables in debug print function to int to silence a compiler warning

stesser committed:
Add correction to result of Goldschmidt	algorithm

The correction is derived from the difference of the input parameter times it inverse.
This value could be used for a final Newton-Raphson step (at the cost of an additional
multiplication), but since we can assume that the input	parameter was very near	to 1.0,
the multiplication would only add non-zero digits beyond the current scale range.

The Goldschmidt	algorithm is an	infinite series	of ever smaller positive values	and
if the series is cut off at some point,	the result is known to be exact	or smaller
than the exact value.

The correction applied is meant	to compensate for the neglected	series elements.
If it is found that there are boundary cases where we over-compensate, then the
Newton-Raphson step could be used to apply a slightly smaller correction.

stesser committed:
Adapt division algorithm to use the updated bc_num_invert()

After a correction is applied to the returned reciprocal to account for the finite
number of elements used in teh Goldschmidt algorithm, the result should be correct
without the final rounding step.

While here also remove the assignment of 1 to a variable that is not used for the
specific case anymore, anyway ...

stesser committed:
Fix width of value printed for base > 17

stesser committed:
Fix printing of fractional parts for obase > 10

stesser committed:
Double correction value applied in bc_num_invert()

It has been found to be too low e.g. when calculating l(256)/l(16) with scale=40.
The result is exact with this increased correction applied.

stesser committed:
Mark a few places where changes might be required with // <se>

The locations with these markers should be checked, although the tests seem to
succeed wíthiut them.

Currently test succeed until "reference.bc2", which returns zeroes for some of
the test arrays. But these marked locations are probably not related to that
test failing.

stesser committed:
Fix length parameter which obviously should be scale not rdx

This change has been performed due to a source code analysis.
No tests failed with the old code (i.e., no test covered this case).

stesser committed:
Add comments to the division based on the Goldschmidt algorithm

stesser committed:
Make shiftLeft return an error for too large shift amounts instead of crashing

The cut-off value is arbitrary, a significantly lower limit could be applied.

stesser committed:
Optimize bc_num_pow10() and remove the unused bc_num_pow10 array

The Goldschmidt algorithm needs larger arguments than supported by the
array.

A larger array could be used, but a disassembly showed, that the array
access was converted to an inlined and rolled out chain of comparisons
and assignments.

gavinhoward commented:
Closed by manually merging.
